### PR TITLE
CONSTRAINT CHECK feature (create table, alter table add column check, alter table drop constraint)

### PR DESCRIPTION
**Pull request by marregui**
Scenario: move logic into the database to avoid having to de-duplicate business constraints among all consumers, e.g. prevent users from inserting x > 15 into a table.

This PR tracks a feature branch.

Refer: https://trello.com/c/lIAaJ26v/85-5-check-constraint-for-create-table

### CODE REVIEW

**Comment by mfussenegger** [comment]
Given that I already reviewed individual commits I think @seut should have another look for sanity checks.

File: docs/appendices/release-notes/unreleased.rst [hunk 1]
```diff
@@ -69,14 +69,18 @@ None
 Changes
 =======
 
+- Add the :ref:`CHECK <check_constraint>` constraint syntax, which specifies
```
Comment (matthijskrul) [inline docs/appendices/release-notes/unreleased.rst:72]: I would not do this / thing, I don't think it adds clarity. A general plural ("values", "columns") is fine.

File: docs/sql/general/constraints.rst [hunk 2]
```diff
@@ -53,6 +53,96 @@ constraint or a table constraint.
 For further details about the meaning of the options see
 :ref:`indices_and_fulltext`.
 
+.. _check_constraint:
+
+``CHECK``
+---------
+
+The CHECK constraint specifies that the value/s of certain column/s must satisfy
```
Comment (matthijskrul) [inline docs/sql/general/constraints.rst]: As above

Comment (marregui) [reply docs/sql/general/constraints.rst]: thank you! both corrected to plural then.

File: sql/src/main/java/io/crate/analyze/CreateTableStatementAnalyzer.java [hunk 3]
```diff
@@ -94,4 +109,39 @@ public AnalyzedCreateTable analyze(CreateTable<Expression> createTable,
...
+                        // Re-frame the column constraint as a table constraint
+                        createTable.tableElements().add(new CheckConstraint<>(
+                            check.name(),
+                            def.ident(),
+                            check.expression(),
+                            check.expressionStr()
+                        ));
+                    }
+                }
+                def.constraints().removeAll(checkColumnConstraints);
```
Comment (seut) [inline sql/src/main/java/io/crate/analyze/CreateTableStatementAnalyzer.java:141]: why is this needed? if needed, could this maybe solved without mutating?

File: sql/src/main/java/io/crate/analyze/DropCheckConstraintAnalyzer.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    public AnalyzedAlterTableDropCheckConstraint analyze(Table<?> table, String name, CoordinatorTxnCtx txnCtx) {
+        DocTableInfo tableInfo = (DocTableInfo) schemas.resolveTableInfo(
+            table.getName(),
+            Operation.ALTER,
+            txnCtx.sessionContext().user(),
+            txnCtx.sessionContext().searchPath());
+        List<CheckConstraint<Symbol>> checkConstraints = tableInfo.checkConstraints();
+        for (int i = 0; i < checkConstraints.size(); i++) {
```
Comment (seut) [inline sql/src/main/java/io/crate/analyze/DropCheckConstraintAnalyzer.java:51]: minor (maybe personal preference): could be also done with simple `for(var checkContraint : checkConstraints)`

File: sql/src/main/java/io/crate/execution/dml/upsert/CheckConstraints.java [hunk 5]
```diff
@@ -71,5 +78,22 @@ public void validate(T values) {
                 throw new IllegalArgumentException("\"" + notNullColumns.get(i) + "\" must not be null");
             }
         }
+        for (int i = 0; i < checkConstraints.size(); i++) {
+            Tuple<? extends Input<?>, CheckConstraint<Symbol>> checkEntry = checkConstraints.get(i);
+            Input<?> checkInput = checkEntry.v1();
+            Boolean value = (Boolean) checkInput.value();
```
Comment (seut) [inline sql/src/main/java/io/crate/execution/dml/upsert/CheckConstraints.java:84]: maybe worth adding an assertion to ensure its a boolean? Is this ensured somewhere else? otherwise it should throw an exception here.

File: sql/src/test/java/io/crate/analyze/CreateAlterTableStatementAnalyzerTest.java [hunk 6]
```diff
@@ -1204,6 +1221,43 @@ private void assertDuplicatePrimaryKey(String stmt) {
...
+        Map<String, Object> mapping = analysis.mapping();
+        Map<String, String> checkConstraints = analysis.analyzedTableElements().getCheckConstraints();
+        assertEquals(checkConstraints.get("check_id_ge_zero"),
+                     Maps.getByPath(mapping, Arrays.asList("_meta", "check_constraints", "check_id_ge_zero")));
+        assertEquals(checkConstraints.get("check_qty_gt_zero"),
+                     Maps.getByPath(mapping, Arrays.asList("_meta", "check_constraints", "check_qty_gt_zero")));
+    }
+
+    @Test
+    public void testAlterTableAddColumnWithCheckConstraint() throws Exception {
```
Comment (seut) [inline sql/src/test/java/io/crate/analyze/CreateAlterTableStatementAnalyzerTest.java:1241]: isn't this a duplicate of https://github.com/crate/crate/pull/9831/files#diff-288536de2bb838c7a805c32bcc63f8ddR397

File: sql/src/test/java/io/crate/execution/dml/upsert/CheckConstraintsTest.java [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    public void setUpExecutor() throws Exception {
+        sqlExecutor = SQLExecutor.builder(clusterService)
+            .addTable("CREATE TABLE t (" +
+                      "    id int," +
+                      "    qty int," +
+                      "    sentinel boolean CONSTRAINT sentinel CHECK(sentinel)," +
+                      "    CONSTRAINT id_is_even CHECK(id % 2 = 0))")
+            .build();
+        docTableInfo = sqlExecutor.resolveTableInfo("t");
+        txnCtx = CoordinatorTxnCtx.systemTransactionContext();
```
Comment (seut) [inline sql/src/test/java/io/crate/execution/dml/upsert/CheckConstraintsTest.java:63]: `sqlExecutor`, `docTableInfo` and  `tnxCtx` can be all a local variables.

File: sql/src/main/java/io/crate/analyze/CreateTableStatementAnalyzer.java (hunk 3 above)
Comment (marregui) [reply sql/src/main/java/io/crate/analyze/CreateTableStatementAnalyzer.java:141]: In this case, what we are doing is treating a check constraint definition at the column level as if it were defined at the table level. During the first pass of the analysis we cannot process check constraints because we do not have references to the columns yet (they are being created).

File: sql/src/main/java/io/crate/execution/dml/upsert/CheckConstraints.java (hunk 5 above)
Comment (marregui) [reply sql/src/main/java/io/crate/execution/dml/upsert/CheckConstraints.java:84]: the grammar makes it be the result of a boolean expression, so it will always be boolean

File: sql/src/main/java/io/crate/analyze/CreateTableStatementAnalyzer.java (hunk 3 above)
Comment (marregui) [reply sql/src/main/java/io/crate/analyze/CreateTableStatementAnalyzer.java:141]: will have a better look at this

File: common/src/main/java/io/crate/common/StringUtils.java [hunk 8]
```diff
@@ -24,11 +24,31 @@
...
+        if (attributes.length > 0) {
+            for (int i = 0; i <= attributes.length / 2; i += 2) {
+                sb.append(Objects.requireNonNull(attributes[i]))
+                    .append("=")
+                    .append(attributes[i + 1])
+                    .append(", ");
+            }
+            sb.setLength(sb.length() - 2);
+        }
+        sb.append("}");
```
Comment (seut) [inline common/src/main/java/io/crate/common/StringUtils.java:48]: I don't think that this complexity is really worth it for simple toString methods. Also, most IDE's have nice toString method generators.

File: sql/src/test/java/io/crate/analyze/CreateAlterTableStatementAnalyzerTest.java [hunk 9]
```diff
@@ -1086,8 +1087,6 @@ public void testCreateTableGeneratedColumnWithMatch() {
 
     @Test
     public void testCreateTableGeneratedColumnBasedOnGeneratedColumn() {
-        expectedException.expect(IllegalArgumentException.class);
-        expectedException.expectMessage("A generated column cannot be based on a generated column");
```
Comment (seut) [inline sql/src/test/java/io/crate/analyze/CreateAlterTableStatementAnalyzerTest.java:1073]: is it expected behaviour that this is now possible?

Comment (mfussenegger) [reply sql/src/test/java/io/crate/analyze/CreateAlterTableStatementAnalyzerTest.java:1073]: I'd be in favour of reverting this behavior change and do it in a separate PR - together with change log entry + dedicated tests (also for the execution layer)

File: sql/src/main/java/io/crate/execution/dml/upsert/CheckConstraints.java (hunk 5 above)
Comment (mfussenegger) [reply sql/src/main/java/io/crate/execution/dml/upsert/CheckConstraints.java:84]: I think it's worth adding an assertion or inline comment that clarifies that

File: sql/src/test/java/io/crate/analyze/CreateAlterTableStatementAnalyzerTest.java (hunk 9 above)
Comment (marregui) [reply sql/src/test/java/io/crate/analyze/CreateAlterTableStatementAnalyzerTest.java:1073]: nope, that was corrected, restoring original behaviour, apologies for the confusion

Comment (marregui) [reply sql/src/test/java/io/crate/analyze/CreateAlterTableStatementAnalyzerTest.java:1073]: Hi @mfussenegger,  just so that I have it correct:

```
 create table foo (
   ts timestamp with time zone,
   day as date_trunc('day', ts),
   date_string as cast(day as string))
```

you are saying that column `date_string` cannot be defined because is defined from day?
with my solution you may. Should I revert that?

Thank you

Comment (marregui) [reply sql/src/test/java/io/crate/analyze/CreateAlterTableStatementAnalyzerTest.java:1073]: I will revert this.

File: sql/src/main/java/io/crate/execution/dml/upsert/CheckConstraints.java (hunk 5 above)
Comment (seut) [reply sql/src/main/java/io/crate/execution/dml/upsert/CheckConstraints.java:84]: Yes lets please add at least a comment, I'd even add an assertion.

Comment (marregui) [reply sql/src/main/java/io/crate/execution/dml/upsert/CheckConstraints.java]: Assertion added.
